src/Aspire.Dashboard/Model/DashboardClient.cs

```diff
@@ -36,23 +35,35 @@ internal sealed class DashboardClient : IDashboardClient
     private ImmutableHashSet<Channel<ResourceViewModelChange>> _outgoingChannels = [];
     private string? _applicationName;
 
+    private const int StateDisabled = -1; // In this state from construction. No dashboard client functions available.
     private const int StateNone = 0;
     private const int StateInitialized = 1;
     private const int StateDisposed = 2;
-    private int _state;
+    private int _state = StateNone;
 
     private readonly GrpcChannel _channel;
     private readonly DashboardService.DashboardServiceClient _client;
 
     private Task? _connection;
 
-    public DashboardClient(ILoggerFactory loggerFactory)
+    public DashboardClient(ILoggerFactory loggerFactory, IEnvironmentVariables environmentVariables)
     {
         _loggerFactory = loggerFactory;
 
         _logger = loggerFactory.CreateLogger<DashboardClient>();
 
-        var address = GetAddressUri(DashboardServiceUrlVariableName, DashboardServiceUrlDefaultValue);
+        var address = environmentVariables.GetUri(DashboardServiceUrlVariableName);
+
+        if (address is null)
+        {
+            _state = StateDisabled;
+            _logger.LogInformation("DOTNET_DASHBOARD_GRPC_ENDPOINT_URL is not specified. Dashboard client services are unavailable.");
+            _cts.Cancel();
+            _whenConnected.TrySetCanceled();
+            _channel = null!;
+            _client = null!;
```

Comment:
I'm not a fan of this. But making it clean would involve another nested type.

Reply:
Won't the dashboard client throw an error when disposed because the channel is null?

Reply:
Maybe they should be nullable and then there is `AssertNotDisabled()` in methods that use these types. It would be annotated so that the compiled knows they're not null if it doesn't throw.

Reply:
I experimented with `MemberNotNull` but didn't like how it looked. For example:

```diff
@@ -126,8 +125,28 @@ GrpcChannel CreateChannel()

     public bool IsEnabled => _state is not StateDisabled;

+    [MemberNotNull(nameof(_channel))]
+    [MemberNotNull(nameof(_client))]
+    [MemberNotNull(nameof(_connection))]
+    private void AssertNotDisabled()
+#pragma warning disable CS8774 // Member must have a non-null value when exiting.
+    {
+        // If we're in disabled state, we're not changing out of it.
+        // We don't need to worry about concurrency for this read.
+        if (_state is StateDisabled)
+        {
+            throw new InvalidOperationException($"{nameof(DashboardClient)} is disabled. Check the {nameof(IsEnabled)} property before calling this.");
+        }
+    }
+#pragma warning restore CS8774 // Member must have a non-null value when exiting.
+
+    [MemberNotNull(nameof(_channel))]
+    [MemberNotNull(nameof(_client))]
+    [MemberNotNull(nameof(_connection))]
     private void EnsureInitialized()
     {
+        AssertNotDisabled();
+
         var priorState = Interlocked.CompareExchange(ref _state, value: StateInitialized, comparand: StateNone);

         if (priorState is StateDisabled)
```

Or, annotating only `EnsureInitialized`:

```diff
@@ -126,6 +125,9 @@ GrpcChannel CreateChannel()

     public bool IsEnabled => _state is not StateDisabled;

+    [MemberNotNull(nameof(_channel))]
+    [MemberNotNull(nameof(_client))]
+    [MemberNotNull(nameof(_connection))]
     private void EnsureInitialized()
     {
         var priorState = Interlocked.CompareExchange(ref _state, value: StateInitialized, comparand: StateNone);
@@ -138,12 +140,16 @@ private void EnsureInitialized()
         if (priorState is not StateNone)
         {
             ObjectDisposedException.ThrowIf(priorState is StateDisposed, this);
+#pragma warning disable CS8774 // Member must have a non-null value when exiting.
             return;
+#pragma warning restore CS8774 // Member must have a non-null value when exiting.
         }

         _connection = Task.Run(() => ConnectAndWatchResourcesAsync(_cts.Token), _cts.Token);

+#pragma warning disable CS8774 // Member must have a non-null value when exiting.
         return;
+#pragma warning restore CS8774 // Member must have a non-null value when exiting.

         async Task ConnectAndWatchResourcesAsync(CancellationToken cancellationToken)
         {
@@ -155,7 +161,7 @@ async Task ConnectAsync()
             {
                 try
                 {
-                    var response = await _client.GetApplicationInformationAsync(new(), cancellationToken: cancellationToken);
+                    var response = await _client!.GetApplicationInformationAsync(new(), cancellationToken: cancellationToken);

                     _applicationName = response.ApplicationName;

@@ -208,7 +214,7 @@ static TimeSpan ExponentialBackOff(int errorCount, double maxSeconds)

                 async Task WatchResourcesAsync()
                 {
-                    var call = _client.WatchResources(new WatchResourcesRequest { IsReconnect = errorCount != 0 }, cancellationToken: cancellationToken);
+                    var call = _client!.WatchResources(new WatchResourcesRequest { IsReconnect = errorCount != 0 }, cancellationToken: cancellationToken);

                     await foreach (var response in call.ResponseStream.ReadAllAsync(cancellationToken: cancellationToken))
                     {
```

I don't like either of those more than what we have now.

> Won't the dashboard client throw an error when disposed because the channel is null?

Yes, thanks I'll fix that. This is what nullability is supposed to catch. I'm going to make the fields nullable and add a three more suppressions, rather than either of the above. If you know a cleaner way, I'd be interested to hear it.